### PR TITLE
[12_4_X] Fix generator interface to Geant4

### PR DESCRIPTION
**Pull request by civanch**
#### PR description:
This is a backport of #40820. See discussions on #39427 and https://indico.cern.ch/event/1242612/contributions/5284691

### CODE REVIEW

**Comment by civanch** [comment]
please test

**Comment by civanch** [comment]
hold

in discussion for #40820 it was proposed that for 12_4_X a modifier is added, because the production is already started. I put this PR on hold and will update soon. 

in discussion

**Comment by civanch** [comment]
@perrotta , @rappoccio , please, confirm that in this case a modifier should be added in 12_4.

**Comment by perrotta** [comment]
> @perrotta , @rappoccio , please, confirm that in this case a modifier should be added in 12_4.

According to what was discussed during the last ORP, given that MC (including SUSY) productions already happened in 12_4: yes, we need a modifier for this backport

**Comment by civanch** [comment]
unhold

**Comment by civanch** [comment]
please test

**Comment by civanch** [comment]
@perrotta , @rappoccio , please confirm that the modifier should be propagated to 12_6 and not the fix from 13_1. I am asking, because in my understanding we have 12_4_X and 13_0_X as MC production releases. Do we need to to fix also 12_6?

**Comment by civanch** [comment]
+1

File: SimG4Core/Generators/src/Generator.cc [hunk 1]
```diff
@@ -470,17 +471,22 @@ void Generator::particleAssignDaughters(G4PrimaryParticle *g4p, HepMC::GenPartic
...
+    int status = (*vpdec)->status();
+    if (verbose > 1)
+      LogDebug("SimG4CoreGenerator::::particleAssignDaughters")
+          << "Assigning a " << (*vpdec)->pdg_id() << " as daughter of a " << vp->pdg_id() << " status=" << status;
+
+    bool checkStatus = fFixG4Primary ? (status > 50 && status < 100) : status > 3;
 
-    if (((*vpdec)->status() == 2 || (*vpdec)->status() > 3) && (*vpdec)->end_vertex() != nullptr) {
+    if ((status == 2 || checkStatus || (status == 23 && std::abs(vp->pdg_id()) == 1000015)) &&
+        (*vpdec)->end_vertex() != nullptr) {
```
Comment (perrotta) [inline SimG4Core/Generators/src/Generator.cc:482]: If `fFixG4Primary` is false, then these lines become
```
    if ((status == 2 || status > 3 || (status == 23 && std::abs(vp->pdg_id()) == 1000015)) &&
        (*vpdec)->end_vertex() != nullptr) {
```
which does not correspond to neither what is in 12_4_X before this PR, nor to the conditions that was there before previous #39440.

Therefore, if `checkStatus` is false, then neither the 12_4_X version previous to Sep 21 is restored, nor the 12_4_X version betweem Sep 21 and now. That looks odd, the default should correspond to the version used for the already started productions.

File: Configuration/ProcessModifiers/python/run3_primaries_for_geant4_2022_cff.py [hunk 2]
```diff
@@ -0,0 +1,6 @@
+import FWCore.ParameterSet.Config as cms
+
+# This modifier introduce the fix of generator/Geant4 interface
+
+run3_g4primaries_2022 = cms.Modifier()
```
Comment (perrotta) [inline Configuration/ProcessModifiers/python/run3_primaries_for_geant4_2022_cff.py:5]: Why "2022" in the name?

File: SimG4Core/Generators/src/Generator.cc (hunk 1 above)
Comment (civanch) [reply SimG4Core/Generators/src/Generator.cc:482]: For me it is the same thing but it is possible to change, so the code will be identical.

File: Configuration/ProcessModifiers/python/run3_primaries_for_geant4_2022_cff.py (hunk 2 above)
Comment (civanch) [reply Configuration/ProcessModifiers/python/run3_primaries_for_geant4_2022_cff.py:5]: This modifier should be applied to 12_4, which is simulation for 2022 data. In 2023 we will use 13_0 and modifier will not be applied. It is possible to remove 2022 but will this removal be OK for operations?

File: SimG4Core/Generators/src/Generator.cc (hunk 1 above)
Comment (civanch) [reply SimG4Core/Generators/src/Generator.cc]: Done

File: Configuration/ProcessModifiers/python/run3_primaries_for_geant4_2022_cff.py (hunk 2 above)
Comment (civanch) [reply Configuration/ProcessModifiers/python/run3_primaries_for_geant4_2022_cff.py]: 2022 is removed

**Comment by civanch** [comment]
please test

**Comment by bsunanda** [comment]
Dear Pratim,

Congratulations. Wish you a successful postdoc period

Sunanda
________________________________

**Comment by civanch** [comment]
+1

**Comment by perrotta** [comment]
+1
- The default reproduces what was in 12_4_X before this PR, i.e. after the merging of previous https://github.com/cms-sw/cmssw/pull/39440
- With the modifier the fix of the generator interface can be enabled

**Pull request merged**